checker/src/org/checkerframework/checker/nullness/KeyForValue.java

```diff
@@ -0,0 +1,113 @@
+package org.checkerframework.checker.nullness;
+
+import com.sun.source.tree.ExpressionTree;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.checker.nullness.qual.KeyFor;
+import org.checkerframework.framework.flow.CFAbstractAnalysis;
+import org.checkerframework.framework.flow.CFAbstractValue;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/**
+ * KeyForValue holds additional information about which maps this value is a key for. This extra
+ * information is required when adding the @KeyFor qualifier to the type is not a refinement of
+ * type. For example,
+ *
+ * <pre>
+ * {@code Map<T, Object> map = ...;}
+ * {@code <T> T method(T param) { }
+ *   if (map.contains(param) {
+ *     {@code @NonNull Object o = map.get(param);}
+ *     return param;
+ *   }
+ * }
+ * }</pre>
+ *
+ * Inside the if statement, {@code param} is a key for "map". This would normally be represented as
+ * {@code @KeyFor("map") T}, but this is not a subtype of {@code T}, so the type cannot be refined.
+ * Instead, the value for {@code param} includes "map" in the list of keyForMaps. This information
+ * is used in {@link KeyForAnnotatedTypeFactory#isKeyForMap(String, ExpressionTree)}.
+ */
+public class KeyForValue extends CFAbstractValue<KeyForValue> {
+    /**
+     * If the underlying type is a type variable or a wildcard, then this is list of maps this for
+     * which this value is a key. Otherwise, it's null.
+     */
+    Set<String> keyForMaps = null;
```

Comment:
Field visibility missing. Make private and add a getter that returns an immutable copy? Or document who is supposed to access?